esmf_regrid/schemes.py

```diff
@@ -1,19 +1,315 @@
+    if cube.coord_system() is None:
+        crs = None
+    else:
+        crs = cube.coord_system().as_cartopy_crs()
+    return GridInfo(
+        lon.points,
+        lat.points,
+        _bounds_cf_to_simple_1d(lon.bounds),
+        _bounds_cf_to_simple_1d(lat.bounds),
+        crs=crs,
+        circular=lon.circular,
+    )
+
+
+def _regrid_along_grid_dims(regridder, data, grid_x_dim, grid_y_dim, mdtol):
+    data = np.moveaxis(data, [grid_x_dim, grid_y_dim], [-1, -2])
+    result = regridder.regrid(data, mdtol=mdtol)
+
+    result = np.moveaxis(result, [-1, -2], [grid_x_dim, grid_y_dim])
+    return result
+
+
+def _create_cube(data, src_cube, grid_dim_x, grid_dim_y, grid_x, grid_y):
+    """
+    Return a new cube for the result of regridding.
+
+    Returned cube represents the result of regridding the source cube
+    onto the new grid.
+    All the metadata and coordinates of the result cube are copied from
+    the source cube, with two exceptions:
+        - Grid dimension coordinates are copied from the grid cube.
+        - Auxiliary coordinates which span the grid dimensions are
+          ignored.
+
+    Parameters
+    ----------
+    data : array
+        The regridded data as an N-dimensional NumPy array.
+    src_cube : cube
+        The source Cube.
+    grid_dim_x : int
+        The dimension of the X coordinate within the source Cube.
+    grid_dim_y : int
+        The dimension of the Y coordinate within the source Cube.
+    grid_x : DimCoord
+        The :class:`iris.coords.DimCoord` for the new grid's X
+        coordinate.
+    grid_y : DimCoord
+        The :class:`iris.coords.DimCoord` for the new grid's Y
+        coordinate.
+
+    Returns
+    -------
+    cube
+        A new iris.cube.Cube instance.
+
+    """
+    new_cube = iris.cube.Cube(data)
+
+    new_cube.add_dim_coord(grid_x, grid_dim_x)
+    new_cube.add_dim_coord(grid_y, grid_dim_y)
+
+    new_cube.metadata = copy.deepcopy(src_cube.metadata)
+
+    def copy_coords(src_coords, add_method):
+        for coord in src_coords:
+            dims = src_cube.coord_dims(coord)
+            if grid_dim_x in dims or grid_dim_y in dims:
+                continue
+            result_coord = coord.copy()
+            # Add result_coord to the owner of add_method.
+            add_method(result_coord, dims)
+
+    copy_coords(src_cube.dim_coords, new_cube.add_dim_coord)
+    copy_coords(src_cube.aux_coords, new_cube.add_aux_coord)
+
+    return new_cube
+
+
+def _regrid_rectilinear_to_rectilinear__prepare(src_grid_cube, tgt_grid_cube):
+    tgt_x, tgt_y = get_xy_dim_coords(tgt_grid_cube)
+    src_x, src_y = get_xy_dim_coords(src_grid_cube)
+
+    grid_x_dim = src_grid_cube.coord_dims(src_x)[0]
+    grid_y_dim = src_grid_cube.coord_dims(src_y)[0]
+
+    srcinfo = _cube_to_GridInfo(src_grid_cube)
+    tgtinfo = _cube_to_GridInfo(tgt_grid_cube)
+
+    regridder = Regridder(srcinfo, tgtinfo)
+
+    regrid_info = (grid_x_dim, grid_y_dim, tgt_x, tgt_y, regridder)
```

Comment:
probably worth considering making `regrid_info` a named tuple if it is going to be passed around between functions.  It will make it much easier to debug if we want to add more values to it in the future

Reply:
That's fair, though it's probably worth noting that we're likely to end up with different regridders which will store different pieces of information. I think we would have to have a named tuple for each scheme.